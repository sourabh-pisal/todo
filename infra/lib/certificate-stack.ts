import { Stack, StackProps } from "aws-cdk-lib";
import {
  Certificate,
  CertificateValidation,
} from "aws-cdk-lib/aws-certificatemanager";
import { Distribution } from "aws-cdk-lib/aws-cloudfront";
import {
  ARecord,
  HostedZone,
  NsRecord,
  RecordTarget,
} from "aws-cdk-lib/aws-route53";
import { CloudFrontTarget } from "aws-cdk-lib/aws-route53-targets";
import { Construct } from "constructs";

export interface CertificateStackProps extends StackProps {
  readonly rootHostedZoneId: string;
  readonly rootHostedZoneName: string;
  readonly subDomainName: string;
}

export class CertificateStack extends Stack {
  private readonly _certificateArn: string;
  private readonly hostedZone: HostedZone;
  constructor(scope: Construct, id: string, props: CertificateStackProps) {
    super(scope, id, props);

    const rootHostedZone = HostedZone.fromHostedZoneAttributes(
      this,
      "RootHostedZone",
      {
        hostedZoneId: props.rootHostedZoneId,
        zoneName: props.rootHostedZoneName,
      },
    );

    this.hostedZone = new HostedZone(this, "HostedZone", {
      zoneName: `${props.subDomainName}.${props.rootHostedZoneName}`,
    });

    new NsRecord(this, "NsRecord", {
      zone: rootHostedZone,
      recordName: props.subDomainName,
      values: this.hostedZone.hostedZoneNameServers!,
    });

    const certificate = new Certificate(this, "Certificate", {
      domainName: `${props.subDomainName}.${props.rootHostedZoneName}`,
      validation: CertificateValidation.fromDns(this.hostedZone),
    });

    this._certificateArn = certificate.certificateArn;
  }

  get certificateArn(): string {
    return this._certificateArn;
  }

  addCloudfrontDelegation(distributionId: string, domainName: string): void {
    new ARecord(this, "AliasRecord", {
      zone: this.hostedZone,
      target: RecordTarget.fromAlias(
        new CloudFrontTarget(
          Distribution.fromDistributionAttributes(this, "Distribution", {
            distributionId: distributionId,
            domainName: domainName,
          }),
        ),
      ),
    });
  }
}
