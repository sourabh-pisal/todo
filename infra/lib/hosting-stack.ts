import { Stack, StackProps } from "aws-cdk-lib";
import { Certificate } from "aws-cdk-lib/aws-certificatemanager";
import {
  AllowedMethods,
  Distribution,
  OriginAccessIdentity,
  ViewerProtocolPolicy,
} from "aws-cdk-lib/aws-cloudfront";
import { S3Origin } from "aws-cdk-lib/aws-cloudfront-origins";
import { Bucket, BucketAccessControl } from "aws-cdk-lib/aws-s3";
import { BucketDeployment, Source } from "aws-cdk-lib/aws-s3-deployment";
import { Construct } from "constructs";
import path = require("path");

const ASSET_PATH = path.resolve(__dirname, "../../todo-app/build");

export interface HostingStackProps extends StackProps {
  readonly certificateArn: string;
  readonly certficateDomainName: string;
}

export class HostingStack extends Stack {
  private readonly distribution: Distribution;

  constructor(scope: Construct, id: string, props: HostingStackProps) {
    super(scope, id, props);

    const hostingBucket = new Bucket(this, "HostingBucket", {
      accessControl: BucketAccessControl.PRIVATE,
    });

    new BucketDeployment(this, "BucketDeployment", {
      destinationBucket: hostingBucket,
      sources: [Source.asset(ASSET_PATH)],
    });

    const originAccessIdentity = new OriginAccessIdentity(
      this,
      "OriginAccessIdentity",
    );

    hostingBucket.grantRead(originAccessIdentity);

    this.distribution = new Distribution(this, "Distribution", {
      defaultRootObject: "index.html",
      defaultBehavior: {
        origin: new S3Origin(hostingBucket, { originAccessIdentity }),
        allowedMethods: AllowedMethods.ALLOW_GET_HEAD,
        compress: true,
        cachedMethods: AllowedMethods.ALLOW_GET_HEAD,
        viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
      },
      certificate: Certificate.fromCertificateArn(
        this,
        "Certificate",
        props.certificateArn,
      ),
      domainNames: [props.certficateDomainName],
    });
  }

  get distributionId(): string {
    return this.distribution.distributionId;
  }

  get distributionDomainName(): string {
    return this.distribution.distributionDomainName;
  }
}
