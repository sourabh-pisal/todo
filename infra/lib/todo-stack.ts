import { Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
// import { Certificate } from "./constructs/certificate";
import {
  Certificate,
  CertificateValidation,
} from "aws-cdk-lib/aws-certificatemanager";
import {
  AllowedMethods,
  Distribution,
  OriginAccessIdentity,
  ViewerProtocolPolicy,
} from "aws-cdk-lib/aws-cloudfront";
import { S3Origin } from "aws-cdk-lib/aws-cloudfront-origins";
import { CfnRecordSet, HostedZone } from "aws-cdk-lib/aws-route53";
import { Bucket, BucketAccessControl } from "aws-cdk-lib/aws-s3";
import { BucketDeployment, Source } from "aws-cdk-lib/aws-s3-deployment";

export interface HostingStackProps extends StackProps {
  readonly hostedZoneId: string;
  readonly domainName: string;
  readonly assetPath: string;
}

export class TodoStack extends Stack {
  constructor(scope: Construct, id: string, props: HostingStackProps) {
    super(scope, id, props);

    const hostedZone = new HostedZone(this, "HostedZone", {
      zoneName: props.domainName,
    });

    const nsRecord = new CfnRecordSet(this, "NsRecord", {
      name: props.domainName,
      type: "NS",
      hostedZoneId: props.hostedZoneId,
      ttl: "1800",
      resourceRecords: hostedZone.hostedZoneNameServers,
    });

    const certificate = new Certificate(this, "Certificate", {
      domainName: props.domainName,
      validation: CertificateValidation.fromDns(hostedZone),
    });

    certificate.node.addDependency(nsRecord);

    const hostingBucket = new Bucket(this, "Bucket", {
      accessControl: BucketAccessControl.PRIVATE,
    });

    new BucketDeployment(this, "BucketDeployment", {
      destinationBucket: hostingBucket,
      sources: [Source.asset(props.assetPath)],
    });

    const originAccessIdentity = new OriginAccessIdentity(
      this,
      "OriginAccessIdentity"
    );

    hostingBucket.grantRead(originAccessIdentity);

    new Distribution(this, "Distribution", {
      defaultRootObject: "index.html",
      defaultBehavior: {
        origin: new S3Origin(hostingBucket, { originAccessIdentity }),
        allowedMethods: AllowedMethods.ALLOW_GET_HEAD,
        compress: true,
        cachedMethods: AllowedMethods.ALLOW_GET_HEAD,
        viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
      },
      certificate: certificate,
      domainNames: [props.domainName],
    });
  }
}
