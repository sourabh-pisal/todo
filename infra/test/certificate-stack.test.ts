import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { CertificateStack } from "../lib/certificate-stack";

test("snapshot", () => {
  const app = new App();
  const certificateStack = new CertificateStack(
    app,
    "TodoAppCertificateStack",
    {
      rootHostedZoneName: "test.com",
      rootHostedZoneId: "hosted-zone-id",
      subDomainName: "sub-domain-name",
    },
  );
  const template = Template.fromStack(certificateStack);
  expect(template.toJSON()).toMatchSnapshot();
});
