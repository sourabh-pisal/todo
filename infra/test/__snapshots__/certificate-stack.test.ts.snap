// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Certificate4E7ABB08": {
      "Properties": {
        "DomainName": "sub-domain-name.test.com",
        "DomainValidationOptions": [
          {
            "DomainName": "sub-domain-name.test.com",
            "HostedZoneId": {
              "Ref": "HostedZoneDB99F866",
            },
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "TodoAppCertificateStack/Certificate",
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
    "HostedZoneDB99F866": {
      "Properties": {
        "Name": "sub-domain-name.test.com.",
      },
      "Type": "AWS::Route53::HostedZone",
    },
    "NsRecordE03782CB": {
      "Properties": {
        "HostedZoneId": "hosted-zone-id",
        "Name": "sub-domain-name.test.com.",
        "ResourceRecords": {
          "Fn::GetAtt": [
            "HostedZoneDB99F866",
            "NameServers",
          ],
        },
        "TTL": "1800",
        "Type": "NS",
      },
      "Type": "AWS::Route53::RecordSet",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
